File: frontend/src/types/security_audit.ts
/**
 * Security Audit & Monitoring Types
 *
 * TypeScript types for security auditing, threat detection,
 * anomaly detection, vulnerability management, and incident management.
 */

// Enums

export enum AuditCategory {
  AUTHENTICATION = 'authentication',
  AUTHORIZATION = 'authorization',
  DATA_ACCESS = 'data_access',
  DATA_MODIFICATION = 'data_modification',
  CONFIGURATION = 'configuration',
  ADMIN_ACTION = 'admin_action',
  SECURITY = 'security',
  SYSTEM = 'system',
}

export enum AuditSeverity {
  DEBUG = 'debug',
  INFO = 'info',
  WARNING = 'warning',
  ERROR = 'error',
  CRITICAL = 'critical',
}

export enum ThreatType {
  BRUTE_FORCE = 'brute_force',
  CREDENTIAL_STUFFING = 'credential_stuffing',
  SQL_INJECTION = 'sql_injection',
  XSS = 'xss',
  CSRF = 'csrf',
  DDoS = 'ddos',
  DATA_EXFILTRATION = 'data_exfiltration',
  PRIVILEGE_ESCALATION = 'privilege_escalation',
  MALWARE = 'malware',
  PHISHING = 'phishing',
  INSIDER_THREAT = 'insider_threat',
  UNKNOWN = 'unknown',
}

export enum ThreatSeverity {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  CRITICAL = 'critical',
}

export enum ThreatStatus {
  DETECTED = 'detected',
  INVESTIGATING = 'investigating',
  CONTAINED = 'contained',
  RESOLVED = 'resolved',
  FALSE_POSITIVE = 'false_positive',
}

export enum IncidentStatus {
  OPEN = 'open',
  INVESTIGATING = 'investigating',
  CONTAINED = 'contained',
  REMEDIATED = 'remediated',
  CLOSED = 'closed',
}

export enum IncidentPriority {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  CRITICAL = 'critical',
}

export enum AnomalyType {
  LOGIN_PATTERN = 'login_pattern',
  ACCESS_PATTERN = 'access_pattern',
  DATA_VOLUME = 'data_volume',
  API_USAGE = 'api_usage',
  GEOGRAPHIC = 'geographic',
  TIME_BASED = 'time_based',
  BEHAVIORAL = 'behavioral',
}

export enum VulnerabilitySeverity {
  NONE = 'none',
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  CRITICAL = 'critical',
}

export enum VulnerabilityStatus {
  OPEN = 'open',
  IN_PROGRESS = 'in_progress',
  REMEDIATED = 'remediated',
  ACCEPTED = 'accepted',
  FALSE_POSITIVE = 'false_positive',
}

// Audit Log Interfaces

export interface AuditLog {
  id: string;
  timestamp: string;
  category: AuditCategory;
  severity: AuditSeverity;
  action: string;
  resource_type: string;
  resource_id?: string;
  user_id?: string;
  user_email?: string;
  organization_id?: string;
  ip_address: string;
  user_agent?: string;
  session_id?: string;
  request_id?: string;
  success: boolean;
  error_code?: string;
  error_message?: string;
  old_value?: Record<string, unknown>;
  new_value?: Record<string, unknown>;
  metadata: Record<string, unknown>;
  geo_location?: Record<string, unknown>;
}

export interface AuditLogCreate {
  category: AuditCategory;
  severity: AuditSeverity;
  action: string;
  resource_type: string;
  resource_id?: string;
  user_id?: string;
  user_email?: string;
  organization_id?: string;
  ip_address: string;
  user_agent?: string;
  session_id?: string;
  success: boolean;
  error_code?: string;
  error_message?: string;
  old_value?: Record<string, unknown>;
  new_value?: Record<string, unknown>;
  metadata?: Record<string, unknown>;
}

export interface AuditLogListResponse {
  logs: AuditLog[];
  total: number;
}

// Threat Intelligence Interfaces

export interface ThreatIndicator {
  id: string;
  threat_type: ThreatType;
  severity: ThreatSeverity;
  indicator_type: string;
  indicator_value: string;
  description?: string;
  source: string;
  confidence: number;
  first_seen: string;
  last_seen: string;
  times_seen: number;
  tags: string[];
  metadata: Record<string, unknown>;
  active: boolean;
  expires_at?: string;
  created_at: string;
}

export interface ThreatIndicatorCreate {
  threat_type: ThreatType;
  severity: ThreatSeverity;
  indicator_type: string;
  indicator_value: string;
  description?: string;
  source?: string;
  confidence?: number;
  tags?: string[];
  expires_at?: string;
}

export interface ThreatDetection {
  id: string;
  threat_type: ThreatType;
  severity: ThreatSeverity;
  status: ThreatStatus;
  title: string;
  description: string;
  source: string;
  confidence: number;
  affected_resources: string[];
  affected_users: string[];
  indicators: string[];
  ip_addresses: string[];
  user_agents: string[];
  organization_id?: string;
  assigned_to?: string;
  detected_at: string;
  acknowledged_at?: string;
  resolved_at?: string;
  resolution_notes?: string;
  related_incidents: string[];
  metadata: Record<string, unknown>;
}

export interface ThreatDetectionListResponse {
  detections: ThreatDetection[];
  total: number;
}

// Anomaly Detection Interfaces

export interface DetectedAnomaly {
  id: string;
  anomaly_type: AnomalyType;
  severity: ThreatSeverity;
  user_id?: string;
  organization_id?: string;
  description: string;
  expected_value?: string;
  actual_value: string;
  deviation_score: number;
  confidence: number;
  ip_address?: string;
  related_events: string[];
  investigated: boolean;
  false_positive: boolean;
  created_at: string;
  investigated_at?: string;
  investigated_by?: string;
}

export interface AnomalyListResponse {
  anomalies: DetectedAnomaly[];
  total: number;
}

// Security Incident Interfaces

export interface SecurityIncident {
  id: string;
  title: string;
  description: string;
  incident_type: ThreatType;
  status: IncidentStatus;
  priority: IncidentPriority;
  severity: ThreatSeverity;
  organization_id: string;
  reported_by: string;
  assigned_to?: string;
  affected_systems: string[];
  affected_users: string[];
  affected_data: string[];
  attack_vectors: string[];
  indicators_of_compromise: string[];
  related_threats: string[];
  timeline: Array<Record<string, unknown>>;
  containment_actions: string[];
  remediation_actions: string[];
  lessons_learned?: string;
  post_mortem_url?: string;
  created_at: string;
  updated_at: string;
  contained_at?: string;
  resolved_at?: string;
  closed_at?: string;
}

export interface SecurityIncidentCreate {
  title: string;
  description: string;
  incident_type: ThreatType;
  priority: IncidentPriority;
  severity: ThreatSeverity;
  organization_id: string;
  affected_systems?: string[];
  affected_users?: string[];
  attack_vectors?: string[];
}

export interface SecurityIncidentUpdate {
  title?: string;
  description?: string;
  status?: IncidentStatus;
  priority?: IncidentPriority;
  assigned_to?: string;
  affected_systems?: string[];
  affected_users?: string[];
  containment_actions?: string[];
  remediation_actions?: string[];
  lessons_learned?: string;
}

export interface SecurityIncidentListResponse {
  incidents: SecurityIncident[];
  total: number;
}

// Vulnerability Interfaces

export interface Vulnerability {
  id: string;
  cve_id?: string;
  title: string;
  description: string;
  severity: VulnerabilitySeverity;
  cvss_score?: number;
  cvss_vector?: string;
  status: VulnerabilityStatus;
  affected_component: string;
  affected_versions: string[];
  organization_id?: string;
  discovered_at: string;
  discovered_by: string;
  assigned_to?: string;
  due_date?: string;
  remediation_steps: string[];
  workaround?: string;
  references: string[];
  exploitability: string;
  patch_available: boolean;
  patch_version?: string;
  verified_at?: string;
  remediated_at?: string;
  created_at: string;
  updated_at: string;
}

export interface VulnerabilityCreate {
  cve_id?: string;
  title: string;
  description: string;
  severity: VulnerabilitySeverity;
  cvss_score?: number;
  affected_component: string;
  affected_versions?: string[];
  organization_id?: string;
  discovered_by: string;
  remediation_steps?: string[];
  references?: string[];
}

export interface VulnerabilityUpdate {
  status?: VulnerabilityStatus;
  assigned_to?: string;
  due_date?: string;
  remediation_steps?: string[];
  workaround?: string;
  patch_available?: boolean;
  patch_version?: string;
}

export interface VulnerabilityListResponse {
  vulnerabilities: Vulnerability[];
  total: number;
}

// Security Metrics Interfaces

export interface SecurityMetrics {
  organization_id: string;
  timestamp: string;
  total_logins_24h: number;
  failed_logins_24h: number;
  unique_users_24h: number;
  mfa_usage_percent: number;
  active_threats: number;
  threats_detected_24h: number;
  threats_resolved_24h: number;
  mean_time_to_detect_hours: number;
  mean_time_to_resolve_hours: number;
  open_incidents: number;
  critical_incidents: number;
  incidents_created_24h: number;
  incidents_resolved_24h: number;
  open_vulnerabilities: number;
  critical_vulnerabilities: number;
  overdue_vulnerabilities: number;
  vulnerabilities_fixed_30d: number;
  anomalies_detected_24h: number;
  false_positive_rate: number;
  audit_log_coverage: number;
  data_classification_coverage: number;
}

export interface SecurityScorecard {
  organization_id: string;
  overall_score: number;
  categories: Record<string, number>;
  strengths: string[];
  weaknesses: string[];
  recommendations: string[];
  calculated_at: string;
}

// Security Alert Interfaces

export interface SecurityAlert {
  id: string;
  alert_type: string;
  severity: ThreatSeverity;
  title: string;
  message: string;
  source: string;
  organization_id?: string;
  related_entity_type?: string;
  related_entity_id?: string;
  acknowledged: boolean;
  acknowledged_by?: string;
  acknowledged_at?: string;
  resolved: boolean;
  resolved_at?: string;
  created_at: string;
}

export interface SecurityAlertListResponse {
  alerts: SecurityAlert[];
  total: number;
}

// Constants

export const AUDIT_CATEGORY_LABELS: Record<AuditCategory, string> = {
  [AuditCategory.AUTHENTICATION]: 'Authentication',
  [AuditCategory.AUTHORIZATION]: 'Authorization',
  [AuditCategory.DATA_ACCESS]: 'Data Access',
  [AuditCategory.DATA_MODIFICATION]: 'Data Modification',
  [AuditCategory.CONFIGURATION]: 'Configuration',
  [AuditCategory.ADMIN_ACTION]: 'Admin Action',
  [AuditCategory.SECURITY]: 'Security',
  [AuditCategory.SYSTEM]: 'System',
};

export const AUDIT_SEVERITY_LABELS: Record<AuditSeverity, string> = {
  [AuditSeverity.DEBUG]: 'Debug',
  [AuditSeverity.INFO]: 'Info',
  [AuditSeverity.WARNING]: 'Warning',
  [AuditSeverity.ERROR]: 'Error',
  [AuditSeverity.CRITICAL]: 'Critical',
};

export const THREAT_TYPE_LABELS: Record<ThreatType, string> = {
  [ThreatType.BRUTE_FORCE]: 'Brute Force Attack',
  [ThreatType.CREDENTIAL_STUFFING]: 'Credential Stuffing',
  [ThreatType.SQL_INJECTION]: 'SQL Injection',
  [ThreatType.XSS]: 'Cross-Site Scripting',
  [ThreatType.CSRF]: 'Cross-Site Request Forgery',
  [ThreatType.DDoS]: 'DDoS Attack',
  [ThreatType.DATA_EXFILTRATION]: 'Data Exfiltration',
  [ThreatType.PRIVILEGE_ESCALATION]: 'Privilege Escalation',
  [ThreatType.MALWARE]: 'Malware',
  [ThreatType.PHISHING]: 'Phishing',
  [ThreatType.INSIDER_THREAT]: 'Insider Threat',
  [ThreatType.UNKNOWN]: 'Unknown',
};

export const THREAT_SEVERITY_LABELS: Record<ThreatSeverity, string> = {
  [ThreatSeverity.LOW]: 'Low',
  [ThreatSeverity.MEDIUM]: 'Medium',
  [ThreatSeverity.HIGH]: 'High',
  [ThreatSeverity.CRITICAL]: 'Critical',
};

export const THREAT_STATUS_LABELS: Record<ThreatStatus, string> = {
  [ThreatStatus.DETECTED]: 'Detected',
  [ThreatStatus.INVESTIGATING]: 'Investigating',
  [ThreatStatus.CONTAINED]: 'Contained',
  [ThreatStatus.RESOLVED]: 'Resolved',
  [ThreatStatus.FALSE_POSITIVE]: 'False Positive',
};

export const INCIDENT_STATUS_LABELS: Record<IncidentStatus, string> = {
  [IncidentStatus.OPEN]: 'Open',
  [IncidentStatus.INVESTIGATING]: 'Investigating',
  [IncidentStatus.CONTAINED]: 'Contained',
  [IncidentStatus.REMEDIATED]: 'Remediated',
  [IncidentStatus.CLOSED]: 'Closed',
};

export const INCIDENT_PRIORITY_LABELS: Record<IncidentPriority, string> = {
  [IncidentPriority.LOW]: 'Low',
  [IncidentPriority.MEDIUM]: 'Medium',
  [IncidentPriority.HIGH]: 'High',
  [IncidentPriority.CRITICAL]: 'Critical',
};

export const VULNERABILITY_SEVERITY_LABELS: Record<VulnerabilitySeverity, string> = {
  [VulnerabilitySeverity.NONE]: 'None',
  [VulnerabilitySeverity.LOW]: 'Low',
  [VulnerabilitySeverity.MEDIUM]: 'Medium',
  [VulnerabilitySeverity.HIGH]: 'High',
  [VulnerabilitySeverity.CRITICAL]: 'Critical',
};

export const VULNERABILITY_STATUS_LABELS: Record<VulnerabilityStatus, string> = {
  [VulnerabilityStatus.OPEN]: 'Open',
  [VulnerabilityStatus.IN_PROGRESS]: 'In Progress',
  [VulnerabilityStatus.REMEDIATED]: 'Remediated',
  [VulnerabilityStatus.ACCEPTED]: 'Risk Accepted',
  [VulnerabilityStatus.FALSE_POSITIVE]: 'False Positive',
};

// Helper Functions

export function getAuditSeverityColor(severity: AuditSeverity): string {
  const colors: Record<AuditSeverity, string> = {
    [AuditSeverity.DEBUG]: 'gray',
    [AuditSeverity.INFO]: 'blue',
    [AuditSeverity.WARNING]: 'yellow',
    [AuditSeverity.ERROR]: 'orange',
    [AuditSeverity.CRITICAL]: 'red',
  };
  return colors[severity] || 'gray';
}

export function getThreatSeverityColor(severity: ThreatSeverity): string {
  const colors: Record<ThreatSeverity, string> = {
    [ThreatSeverity.LOW]: 'blue',
    [ThreatSeverity.MEDIUM]: 'yellow',
    [ThreatSeverity.HIGH]: 'orange',
    [ThreatSeverity.CRITICAL]: 'red',
  };
  return colors[severity] || 'gray';
}

export function getThreatStatusColor(status: ThreatStatus): string {
  const colors: Record<ThreatStatus, string> = {
    [ThreatStatus.DETECTED]: 'red',
    [ThreatStatus.INVESTIGATING]: 'yellow',
    [ThreatStatus.CONTAINED]: 'blue',
    [ThreatStatus.RESOLVED]: 'green',
    [ThreatStatus.FALSE_POSITIVE]: 'gray',
  };
  return colors[status] || 'gray';
}

export function getIncidentStatusColor(status: IncidentStatus): string {
  const colors: Record<IncidentStatus, string> = {
    [IncidentStatus.OPEN]: 'red',
    [IncidentStatus.INVESTIGATING]: 'yellow',
    [IncidentStatus.CONTAINED]: 'blue',
    [IncidentStatus.REMEDIATED]: 'green',
    [IncidentStatus.CLOSED]: 'gray',
  };
  return colors[status] || 'gray';
}

export function getVulnerabilitySeverityColor(severity: VulnerabilitySeverity): string {
  const colors: Record<VulnerabilitySeverity, string> = {
    [VulnerabilitySeverity.NONE]: 'gray',
    [VulnerabilitySeverity.LOW]: 'blue',
    [VulnerabilitySeverity.MEDIUM]: 'yellow',
    [VulnerabilitySeverity.HIGH]: 'orange',
    [VulnerabilitySeverity.CRITICAL]: 'red',
  };
  return colors[severity] || 'gray';
}

export function formatConfidence(confidence: number): string {
  return `${Math.round(confidence * 100)}%`;
}

export function formatDeviationScore(score: number): string {
  return `${score.toFixed(1)}σ`;
}

export function getSecurityScoreColor(score: number): string {
  if (score >= 80) return 'green';
  if (score >= 60) return 'blue';
  if (score >= 40) return 'yellow';
  if (score >= 20) return 'orange';
  return 'red';
}

export function getSecurityScoreLabel(score: number): string {
  if (score >= 80) return 'Excellent';
  if (score >= 60) return 'Good';
  if (score >= 40) return 'Fair';
  if (score >= 20) return 'Poor';
  return 'Critical';
}

export function isVulnerabilityOverdue(vulnerability: Vulnerability): boolean {
  if (!vulnerability.due_date) return false;
  if (vulnerability.status === VulnerabilityStatus.REMEDIATED) return false;
  return new Date(vulnerability.due_date) < new Date();
}

export function formatCVSSScore(score?: number): string {
  if (score === undefined || score === null) return 'N/A';
  return score.toFixed(1);
}

export function getCVSSScoreColor(score?: number): string {
  if (score === undefined || score === null) return 'gray';
  if (score >= 9.0) return 'red';
  if (score >= 7.0) return 'orange';
  if (score >= 4.0) return 'yellow';
  if (score > 0) return 'blue';
  return 'gray';
}

export function getIncidentDuration(incident: SecurityIncident): string {
  const start = new Date(incident.created_at);
  const end = incident.closed_at ? new Date(incident.closed_at) : new Date();
  const diffMs = end.getTime() - start.getTime();
  const hours = Math.floor(diffMs / (1000 * 60 * 60));
  const days = Math.floor(hours / 24);

  if (days > 0) return `${days}d ${hours % 24}h`;
  return `${hours}h`;
}

export function getThreatIcon(type: ThreatType): string {
  const icons: Record<ThreatType, string> = {
    [ThreatType.BRUTE_FORCE]: 'key',
    [ThreatType.CREDENTIAL_STUFFING]: 'users',
    [ThreatType.SQL_INJECTION]: 'database',
    [ThreatType.XSS]: 'code',
    [ThreatType.CSRF]: 'link',
    [ThreatType.DDoS]: 'cloud-off',
    [ThreatType.DATA_EXFILTRATION]: 'download-cloud',
    [ThreatType.PRIVILEGE_ESCALATION]: 'arrow-up',
    [ThreatType.MALWARE]: 'alert-triangle',
    [ThreatType.PHISHING]: 'mail',
    [ThreatType.INSIDER_THREAT]: 'user-x',
    [ThreatType.UNKNOWN]: 'help-circle',
  };
  return icons[type] || 'alert-circle';
}

File: backend/app/api/v1/endpoints/security_audit.py
"""
Security Audit & Monitoring API Endpoints

REST API for security auditing, threat detection, anomaly detection,
vulnerability management, and security incident management.
"""

from fastapi import APIRouter, HTTPException, Query
from typing import Optional
from datetime import datetime

from app.schemas.security_audit import (
    AuditLog, AuditLogCreate, AuditLogListResponse,
    AuditCategory, AuditSeverity,
    ThreatIndicator, ThreatIndicatorCreate,
    ThreatDetection, ThreatDetectionListResponse,
    ThreatType, ThreatSeverity, ThreatStatus,
    DetectedAnomaly, AnomalyListResponse, AnomalyType,
    SecurityIncident, SecurityIncidentCreate, SecurityIncidentUpdate,
    SecurityIncidentListResponse, IncidentStatus, IncidentPriority,
    Vulnerability, VulnerabilityCreate, VulnerabilityUpdate,
    VulnerabilityListResponse, VulnerabilitySeverity, VulnerabilityStatus,
    SecurityMetrics, SecurityScorecard,
    SecurityAlert, SecurityAlertListResponse,
)
from app.services.security_audit_service import security_audit_service

router = APIRouter()


# Audit Logs

@router.post("/audit-logs", response_model=AuditLog, tags=["audit-logs"])
async def create_audit_log(data: AuditLogCreate):
    """Create an audit log entry."""
    return security_audit_service.create_audit_log(data)


@router.get("/audit-logs", response_model=AuditLogListResponse, tags=["audit-logs"])
async def list_audit_logs(
    category: Optional[AuditCategory] = Query(None),
    severity: Optional[AuditSeverity] = Query(None),
    action: Optional[str] = Query(None),
    resource_type: Optional[str] = Query(None),
    user_id: Optional[str] = Query(None),
    organization_id: Optional[str] = Query(None),
    ip_address: Optional[str] = Query(None),
    success: Optional[bool] = Query(None),
    start_time: Optional[datetime] = Query(None),
    end_time: Optional[datetime] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
):
    """List audit logs with filtering."""
    from app.schemas.security_audit import AuditLogFilter
    filter = AuditLogFilter(
        category=category,
        severity=severity,
        action=action,
        resource_type=resource_type,
        user_id=user_id,
        organization_id=organization_id,
        ip_address=ip_address,
        success=success,
        start_time=start_time,
        end_time=end_time,
    )
    return security_audit_service.list_audit_logs(filter, skip, limit)


@router.get("/audit-logs/{log_id}", response_model=AuditLog, tags=["audit-logs"])
async def get_audit_log(log_id: str):
    """Get audit log by ID."""
    log = security_audit_service.get_audit_log(log_id)
    if not log:
        raise HTTPException(status_code=404, detail="Audit log not found")
    return log


# Threat Intelligence

@router.post("/threat-indicators", response_model=ThreatIndicator, tags=["threat-intelligence"])
async def create_threat_indicator(data: ThreatIndicatorCreate):
    """Create threat indicator."""
    return security_audit_service.create_threat_indicator(data)


@router.get("/threat-indicators", tags=["threat-intelligence"])
async def list_threat_indicators(
    threat_type: Optional[ThreatType] = Query(None),
    active_only: bool = Query(True),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
):
    """List threat indicators."""
    indicators = security_audit_service.list_threat_indicators(
        threat_type, active_only, skip, limit
    )
    return {"indicators": indicators, "total": len(indicators)}


@router.get("/threat-indicators/{indicator_id}", response_model=ThreatIndicator, tags=["threat-intelligence"])
async def get_threat_indicator(indicator_id: str):
    """Get threat indicator by ID."""
    indicator = security_audit_service.get_threat_indicator(indicator_id)
    if not indicator:
        raise HTTPException(status_code=404, detail="Threat indicator not found")
    return indicator


@router.post("/threat-indicators/check", tags=["threat-intelligence"])
async def check_indicator(
    indicator_type: str = Query(..., description="Type: ip, domain, hash, email"),
    value: str = Query(...),
):
    """Check if value matches any threat indicator."""
    indicator = security_audit_service.check_indicator(indicator_type, value)
    return {
        "matched": indicator is not None,
        "indicator": indicator,
    }


# Threat Detection

@router.get("/threats", response_model=ThreatDetectionListResponse, tags=["threat-detection"])
async def list_threat_detections(
    organization_id: Optional[str] = Query(None),
    status: Optional[ThreatStatus] = Query(None),
    severity: Optional[ThreatSeverity] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
):
    """List threat detections."""
    return security_audit_service.list_threat_detections(
        organization_id, status, severity, skip, limit
    )


@router.get("/threats/{detection_id}", response_model=ThreatDetection, tags=["threat-detection"])
async def get_threat_detection(detection_id: str):
    """Get threat detection by ID."""
    detection = security_audit_service.get_threat_detection(detection_id)
    if not detection:
        raise HTTPException(status_code=404, detail="Threat detection not found")
    return detection


@router.patch("/threats/{detection_id}", response_model=ThreatDetection, tags=["threat-detection"])
async def update_threat_detection(
    detection_id: str,
    status: Optional[ThreatStatus] = Query(None),
    assigned_to: Optional[str] = Query(None),
    resolution_notes: Optional[str] = Query(None),
):
    """Update threat detection."""
    detection = security_audit_service.update_threat_detection(
        detection_id, status, assigned_to, resolution_notes
    )
    if not detection:
        raise HTTPException(status_code=404, detail="Threat detection not found")
    return detection


# Anomaly Detection

@router.post("/anomalies", response_model=DetectedAnomaly, tags=["anomaly-detection"])
async def detect_anomaly(
    anomaly_type: AnomalyType = Query(...),
    description: str = Query(...),
    actual_value: str = Query(...),
    deviation_score: float = Query(..., ge=0),
    user_id: Optional[str] = Query(None),
    organization_id: Optional[str] = Query(None),
    ip_address: Optional[str] = Query(None),
    expected_value: Optional[str] = Query(None),
):
    """Record detected anomaly."""
    return security_audit_service.detect_anomaly(
        anomaly_type=anomaly_type,
        description=description,
        actual_value=actual_value,
        deviation_score=deviation_score,
        user_id=user_id,
        organization_id=organization_id,
        ip_address=ip_address,
        expected_value=expected_value,
    )


@router.get("/anomalies", response_model=AnomalyListResponse, tags=["anomaly-detection"])
async def list_anomalies(
    organization_id: Optional[str] = Query(None),
    user_id: Optional[str] = Query(None),
    anomaly_type: Optional[AnomalyType] = Query(None),
    investigated: Optional[bool] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
):
    """List detected anomalies."""
    return security_audit_service.list_anomalies(
        organization_id, user_id, anomaly_type, investigated, skip, limit
    )


@router.post("/anomalies/{anomaly_id}/investigate", response_model=DetectedAnomaly, tags=["anomaly-detection"])
async def investigate_anomaly(
    anomaly_id: str,
    investigated_by: str = Query(...),
    false_positive: bool = Query(False),
):
    """Mark anomaly as investigated."""
    anomaly = security_audit_service.investigate_anomaly(
        anomaly_id, investigated_by, false_positive
    )
    if not anomaly:
        raise HTTPException(status_code=404, detail="Anomaly not found")
    return anomaly


# Security Incidents

@router.post("/incidents", response_model=SecurityIncident, tags=["security-incidents"])
async def create_incident(
    data: SecurityIncidentCreate,
    reported_by: str = Query(...),
):
    """Create security incident."""
    return security_audit_service.create_incident(data, reported_by)


@router.get("/incidents", response_model=SecurityIncidentListResponse, tags=["security-incidents"])
async def list_incidents(
    organization_id: Optional[str] = Query(None),
    status: Optional[IncidentStatus] = Query(None),
    priority: Optional[IncidentPriority] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
):
    """List security incidents."""
    return security_audit_service.list_incidents(
        organization_id, status, priority, skip, limit
    )


@router.get("/incidents/{incident_id}", response_model=SecurityIncident, tags=["security-incidents"])
async def get_incident(incident_id: str):
    """Get security incident by ID."""
    incident = security_audit_service.get_incident(incident_id)
    if not incident:
        raise HTTPException(status_code=404, detail="Incident not found")
    return incident


@router.patch("/incidents/{incident_id}", response_model=SecurityIncident, tags=["security-incidents"])
async def update_incident(
    incident_id: str,
    data: SecurityIncidentUpdate,
    updated_by: str = Query(...),
):
    """Update security incident."""
    incident = security_audit_service.update_incident(incident_id, data, updated_by)
    if not incident:
        raise HTTPException(status_code=404, detail="Incident not found")
    return incident


@router.post("/incidents/{incident_id}/notes", response_model=SecurityIncident, tags=["security-incidents"])
async def add_incident_note(
    incident_id: str,
    note: str = Query(...),
    user: str = Query(...),
):
    """Add note to incident timeline."""
    incident = security_audit_service.add_incident_note(incident_id, note, user)
    if not incident:
        raise HTTPException(status_code=404, detail="Incident not found")
    return incident


# Vulnerability Management

@router.post("/vulnerabilities", response_model=Vulnerability, tags=["vulnerability-management"])
async def create_vulnerability(data: VulnerabilityCreate):
    """Create vulnerability record."""
    return security_audit_service.create_vulnerability(data)


@router.get("/vulnerabilities", response_model=VulnerabilityListResponse, tags=["vulnerability-management"])
async def list_vulnerabilities(
    organization_id: Optional[str] = Query(None),
    status: Optional[VulnerabilityStatus] = Query(None),
    severity: Optional[VulnerabilitySeverity] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
):
    """List vulnerabilities."""
    return security_audit_service.list_vulnerabilities(
        organization_id, status, severity, skip, limit
    )


@router.get("/vulnerabilities/{vuln_id}", response_model=Vulnerability, tags=["vulnerability-management"])
async def get_vulnerability(vuln_id: str):
    """Get vulnerability by ID."""
    vulnerability = security_audit_service.get_vulnerability(vuln_id)
    if not vulnerability:
        raise HTTPException(status_code=404, detail="Vulnerability not found")
    return vulnerability


@router.patch("/vulnerabilities/{vuln_id}", response_model=Vulnerability, tags=["vulnerability-management"])
async def update_vulnerability(vuln_id: str, data: VulnerabilityUpdate):
    """Update vulnerability."""
    vulnerability = security_audit_service.update_vulnerability(vuln_id, data)
    if not vulnerability:
        raise HTTPException(status_code=404, detail="Vulnerability not found")
    return vulnerability


# Security Alerts

@router.get("/alerts", response_model=SecurityAlertListResponse, tags=["security-alerts"])
async def list_alerts(
    organization_id: Optional[str] = Query(None),
    severity: Optional[ThreatSeverity] = Query(None),
    acknowledged: Optional[bool] = Query(None),
    resolved: Optional[bool] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100),
):
    """List security alerts."""
    return security_audit_service.list_alerts(
        organization_id, severity, acknowledged, resolved, skip, limit
    )


@router.post("/alerts/{alert_id}/acknowledge", response_model=SecurityAlert, tags=["security-alerts"])
async def acknowledge_alert(
    alert_id: str,
    acknowledged_by: str = Query(...),
):
    """Acknowledge security alert."""
    alert = security_audit_service.acknowledge_alert(alert_id, acknowledged_by)
    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")
    return alert


@router.post("/alerts/{alert_id}/resolve", response_model=SecurityAlert, tags=["security-alerts"])
async def resolve_alert(alert_id: str):
    """Resolve security alert."""
    alert = security_audit_service.resolve_alert(alert_id)
    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")
    return alert


# Security Metrics

@router.get("/organizations/{org_id}/metrics", response_model=SecurityMetrics, tags=["security-metrics"])
async def get_security_metrics(org_id: str):
    """Get security metrics for organization."""
    return security_audit_service.get_security_metrics(org_id)


@router.get("/organizations/{org_id}/scorecard", response_model=SecurityScorecard, tags=["security-metrics"])
async def get_security_scorecard(org_id: str):
    """Get security scorecard for organization."""
    return security_audit_service.get_security_scorecard(org_id)
